File: ase-backend/src/main/java/de/corvinrose/ase/controller/ShaderController.java
package de.corvinrose.ase.controller;

import de.corvinrose.ase.model.Shader;
import de.corvinrose.ase.service.ShaderService;
import java.util.List;
import java.util.UUID;
import lombok.AllArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.*;

@AllArgsConstructor
@RestController
@RequestMapping("/shader")
public class ShaderController {

    private final ShaderService shaderService;

    @GetMapping
    public ResponseEntity<List<Shader>> getAllShaders() {
        List<Shader> shaders = shaderService.findAllShaders();
        return new ResponseEntity<>(shaders, HttpStatus.OK);
    }

    @GetMapping("/{id}")
    public ResponseEntity<Shader> getShaderById(@PathVariable("id") UUID id) {
        Shader shader =
                shaderService
                        .findShaderById(id)
                        .orElseThrow(
                                () -> {
                                    throw new IllegalArgumentException(
                                            "No Shader with id " + id.toString() + " exists");
                                });
        return new ResponseEntity<>(shader, HttpStatus.OK);
    }

    @PostMapping("/add")
    public ResponseEntity<Shader> addShader(@RequestBody Shader shader) {
        Shader newShader = shaderService.addShader(shader);
        return new ResponseEntity<>(newShader, HttpStatus.CREATED);
    }

    @PutMapping("/update")
    public ResponseEntity<Shader> updateShader(@RequestBody Shader shader) {
        Shader updatedShader = shaderService.updateShader(shader);
        return new ResponseEntity<>(updatedShader, HttpStatus.OK);
    }

    @PatchMapping("/update")
    public ResponseEntity<Shader> patchUser(@RequestBody Shader shader) {
        return updateShader(shader);
    }

    @Transactional
    @DeleteMapping("/delete/{id}")
    public ResponseEntity<Shader> deleteShader(@PathVariable("id") UUID id) {
        shaderService.deleteShader(id);
        return new ResponseEntity<>(HttpStatus.OK);
    }
}

File: backend/Port/Web/Controller/ShaderController.php
<?php

namespace Port\Web\Controller;

use Core\App\Provider\BaseController;
use Core\App\Service\AuthService;
use Core\App\Service\ShaderService;
use Core\Domain\Model\Routing\ContentType;
use Core\Domain\Model\Routing\HttpStatus;
use Core\Domain\Model\Routing\ResponseEntity;
use Core\Domain\Model\ShaderModel;
use Exception;

class ShaderController extends BaseController {
    private ShaderService $shaderService;
    private AuthService $authService;

    public function __construct() {
        $this->shaderService = new ShaderService();
        $this->authService = new AuthService();
    }

    /**
     * @Get /v1/shader
     * @throws Exception
     */
    public function getAllShaders(): ResponseEntity {
        return new ResponseEntity($this->shaderService->findAllShaders(), HttpStatus::OK());
    }

    /**
     * @Get /v1/shader/{id}
     * @throws Exception
     */
    public function getShaderById(string $id): ResponseEntity {
        return new ResponseEntity($this->shaderService->findShaderById($id), HttpStatus::OK());
    }

    /**
     * @Post /v1/shader/add
     * @throws Exception
     */
    public function addShader(): ResponseEntity {
        $authStatus = $this->authService->authStatus();
        $requestBody = file_get_contents("php://input");
        $shader = ShaderModel::fromJsonObject(json_decode($requestBody));
        return new ResponseEntity($this->shaderService->addShader($shader), HttpStatus::CREATED());
    }

    /**
     * @Put /v1/shader/update
     * @throws Exception
     */
    public function updateShader(): ResponseEntity {
        $authStatus = $this->authService->authStatus();
        $requestBody = file_get_contents("php://input");
        $shader = ShaderModel::fromJsonObject(json_decode($requestBody));
        $user = $this->authService->findUserByEmail($authStatus->userName);
        if ($this->shaderService->findShaderById($shader->id)->authorId !== $user->id) {
            return $this->authService->unauthorized();
        }
        return new ResponseEntity($this->shaderService->updateShader($shader), HttpStatus::OK());
    }

    /**
     * @Delete /v1/shader/delete/{id}
     * @throws Exception
     */
    public function deleteShader(string $id): ResponseEntity {
        $authStatus = $this->authService->authStatus();
        $shader = $this->shaderService->findShaderById($id);
        $user = $this->authService->findUserByEmail($authStatus->userName);
        if ($shader->authorId !== $user->id) {
            return $this->authService->unauthorized();
        }
        return new ResponseEntity($this->shaderService->deleteShader($id), HttpStatus::OK());
    }

    /**
     * @Get /v1/shader/{id}/image
     */
    public function getShaderImage(string $id): ResponseEntity {
        try {
            $img = $this->shaderService->findShaderById($id)->previewImg;
        } catch (Exception $e) {
            $img = "R0lGODlhAQABAIAAAAUEBAAAACwAAAAAAQABAAACAkQBADs=";
        }
        $imageData = str_replace("data:image/jpeg;base64,", "", $img);
        $imageData = base64_decode($imageData);
        return new ResponseEntity($imageData, HttpStatus::OK(), ContentType::IMAGE_JPEG);
    }
}
